### PR TITLE
Fixed Ice Spinner, refactored battle terrain text string tables and ditched EFFECT_REMOVE_TERRAIN

### LINKED ISSUE #3245: Ice Spinner SoftResets Game
### Description

https://github.com/rh-hideout/pokeemerald-expansion/assets/96850956/bcfe7d68-48f6-4055-acf6-04e63efe6445

The Move **Ice Spinner** works correctly, however after its effect, it soft resets the game as seen in the video above. The treecko is only modified to test the move out, as well as the Poochyena only being modified to test Ice Spinner's functionality.

### Version

1.5.2 (Default)

### Upcoming Version

_No response_

### Discord contact info

@denebony

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
This PR fixes #3245.
Ice Spinner was making the game crash because it was trying to read a `STRINGID_TERRAINREMOVED` constant from inside `gTerrainStringIds` that wasn't linked to any battle string at `gBattleStringsTable`.

After looking into the code that uses that string table a bit more deeply, I noticed there was room to clean things up a bit, so I decided to unite that table, with the table that contained the strings printed when the code needs to remove an existing terrain.
Another reason why I did that is to remain consistent with the effect of the `setterrain` battle script command, which handled 2 different situations; move effects that could set a terrain and move effects that could end a terrain.
*If the battle script command handles setting and removing a terrain, why shouldn't the string table for the terrains also handle both situations?*

On that note, I renamed `setterrain` to `setremoveterrain` because that reflects its effect more accurately.

## **Discord contact info**
lunos4026

### CODE REVIEW

**Comment by LOuroboros** [comment]
I updated the PR. There's exactly 0 need to have a separate battle move effect to remove terrains when `EFFECT_SET_REMOVE_TERRAIN` can handle things just fine, so I took this chance to remove it.

**Comment by LOuroboros** [comment]
Putting this in Draft Mode until I fix the tests.

**Review by DenEbony — APPROVED** [review]
From personal testing, this does solve the problem of the original post. Thanks much!


https://github.com/rh-hideout/pokeemerald-expansion/assets/96850956/ba08f6a9-c008-459f-b44d-c92a5612396b

**Comment by LOuroboros** [comment]
Turns out `EFFECT_REMOVE_TERRAIN` is not that pointless. We have to keep it because Steel Roller, unlike Ice Spinner, actually fails if there is no terrain active. Quite the Game Freak moment.

I reverted the commit and force pushed, since there weren't any reviewed made thus far.

**Comment by LOuroboros** [comment]
Converting to draft for now.

I'd like to hear some opinions about an idea I had after that comment I wrote 2 hours ago.

Basically, I ended up adding an argument value to handle Steel Roller's flavor of a terrain-removing move effect and I also took the chance to add some constant labels for the 3 possible argument values that `EFFECT_HIT_SET_REMOVE_TERRAIN` can use , but I'd like to hear some opinions before pushing it.

Here's the diff: https://pastebin.com/0fNyLzf3

**Comment by LOuroboros** [comment]
I didn't feel like I was going to get any feedback soon, so I just went ahead and pushed those changes.

Ready for review.

**Comment by LOuroboros** [comment]
Since they're related to the battle scripts that I'm touching over here, I decided to take the chance and get rid of a few `various`es involved, namely `VARIOUS_SET_REMOVE_TERRAIN` and `VARIOUS_JUMP_IF_TERRAIN_AFFECTED`.
Both of them have been turned into nice and comfy `callnative`s.

I also merged `HandleTerrainMove` into the shiny new `BS_SetRemoveTerrain` because there was no reason to.
The function couldn't have been called from outside of `src/battle_script_commands.c` because it deals with battle instructions, and it was only ever used for `VARIOUS_SET_REMOVE_TERRAIN`.

Re-ready for review.

**Comment by AsparagusEduardo** [comment]
I'm not too sure about merging the effect of removing terrain with the one that sets it.

I need to take a better look at the approach, but for now, please write the tests.

**Comment by LOuroboros** [comment]
> I'm not too sure about merging the effect of removing terrain with the one that sets it.

Just so you know, `EFFECT_HIT_SET_REMOVE_TERRAIN` already removed terrains as a part of Ice Spinner's effect.

What I merged into it was Steel Roller's own variation of it.

> I need to take a better look at the approach, but for now, please write the tests.

Err... I don't want to do that myself, but I'll think about it later.

File: test/battle/move_effect/hit_set_remove_terrain.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gBattleMoves[MOVE_ELECTRIC_TERRAIN].effect == MOVE_ELECTRIC_TERRAIN);
+    ASSUME(gBattleMoves[MOVE_PSYCHIC_TERRAIN].effect == EFFECT_PSYCHIC_TERRAIN);
+    ASSUME(gBattleMoves[MOVE_GRASSY_TERRAIN].effect == EFFECT_GRASSY_TERRAIN);
+    ASSUME(gBattleMoves[MOVE_MISTY_TERRAIN].effect == EFFECT_MISTY_TERRAIN);
+    ASSUME(gBattleMoves[MOVE_STEEL_ROLLER].effect == EFFECT_RECOIL_IF_MISS);
+    ASSUME(gBattleMoves[MOVE_ICE_SPINNER].effect == EFFECT_RECOIL_IF_MISS);
```
Comment (AsparagusEduardo) [inline test/battle/move_effect/hit_set_remove_terrain.c:11]: Neither of them are `EFFECT_HIT_SET_REMOVE_TERRAIN` ._.

File: test/battle/move_effect/hit_set_remove_terrain.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+SINGLE_BATTLE_TEST("Steel Roller fails if there is no terrain on the field")
+{
+    GIVEN {
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_STEEL_ROLLER); }
+    } SCENE {
+        MESSAGE("But it failed!");
```
Comment (AsparagusEduardo) [inline test/battle/move_effect/hit_set_remove_terrain.c:68]:
```suggestion
        NOT ANIMATION(ANIM_TYPE_MOVE, MOVE_STEEL_ROLLER, player);
        MESSAGE("But it failed!");
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: test/battle/move_effect/hit_set_remove_terrain.c (hunk 1 above)
Comment (AlexOn1ine) [reply test/battle/move_effect/hit_set_remove_terrain.c:11]: my bad, sorry
```
    ASSUME(gBattleMoves[MOVE_STEEL_ROLLER].effect == EFFECT_HIT_SET_REMOVE_TERRAIN);
    ASSUME(gBattleMoves[MOVE_ICE_SPINNER].effect == EFFECT_HIT_SET_REMOVE_TERRAIN);
```

Comment (AsparagusEduardo) [reply test/battle/move_effect/hit_set_remove_terrain.c:11]: I know, it's an issue with `ASSUMPTIONS`: #3043

Comment (AlexOn1ine) [reply test/battle/move_effect/hit_set_remove_terrain.c:11]: so basically assumptions currently don't matter?

Comment (AsparagusEduardo) [reply test/battle/move_effect/hit_set_remove_terrain.c:11]: They should still be fixed

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**